esp-hal/src/rsa/mod.rs

```diff
@@ -47,29 +50,43 @@ pub struct Rsa<'d, DM: crate::Mode> {
     phantom: PhantomData<DM>,
 }
 
-impl<'d> Rsa<'d, crate::Blocking> {
+impl<'d> Rsa<'d, Blocking> {
     /// Create a new instance in [crate::Blocking] mode.
     ///
     /// Optionally an interrupt handler can be bound.
     pub fn new(rsa: impl Peripheral<P = RSA> + 'd) -> Self {
         Self::new_internal(rsa)
     }
+
+    /// Reconfigures the RSA driver to operate in asynchronous mode.
+    pub fn into_async(mut self) -> Rsa<'d, Async> {
+        self.set_interrupt_handler(asynch::rsa_interrupt_handler);
+        Rsa {
+            rsa: self.rsa,
+            phantom: PhantomData,
+        }
+    }
 }
 
-impl<'d> crate::private::Sealed for Rsa<'d, crate::Blocking> {}
+impl crate::private::Sealed for Rsa<'_, Blocking> {}
 
-impl<'d> InterruptConfigurable for Rsa<'d, crate::Blocking> {
+impl InterruptConfigurable for Rsa<'_, Blocking> {
     fn set_interrupt_handler(&mut self, handler: InterruptHandler) {
-        self.internal_set_interrupt_handler(handler);
+        unsafe { crate::interrupt::bind_interrupt(Interrupt::RSA, handler.handler()) };
+        unwrap!(crate::interrupt::enable(Interrupt::RSA, handler.priority()));
     }
 }
 
-impl<'d> Rsa<'d, crate::Async> {
+impl<'d> Rsa<'d, Async> {
     /// Create a new instance in [crate::Blocking] mode.
-    pub fn new_async(rsa: impl Peripheral<P = RSA> + 'd) -> Self {
-        let mut this = Self::new_internal(rsa);
-        this.internal_set_interrupt_handler(asynch::rsa_interrupt_handler);
-        this
+    pub fn into_blocking(self) -> Rsa<'d, Blocking> {
+        crate::interrupt::disable(Cpu::ProCpu, Interrupt::RSA);
+        #[cfg(multi_core)]
+        crate::interrupt::disable(Cpu::AppCpu, Interrupt::RSA);
```

Comment:
Why disable on both cores? Just the current one should suffice, no?

Reply:
The drivers are currently Send, so it's possible to drop one on the other core. We don't assume where a driver was created (except to bind its interrupts in the first place, which is problematic as it is).

Reply:
Considering the interrupt handlers are not written to handle cross core setups, the async drivers should be marked at `!Send`. If folks wants to move the drivers, they should do so in blocking mode.

Reply:
> Considering the interrupt handlers are not written to handle cross core setups, the async drivers should be marked at !Send. If folks wants to move the drivers, they should do so in blocking mode.

This is a nice idea and I think resolves https://github.com/esp-rs/esp-hal/issues/1525 properly.

Reply:
I'll do this separately, there are weird edge cases with split peripherals.